zerver/lib/push_notifications.py

```diff
@@ -256,7 +256,7 @@ def get_info_from_apns_result(
         logger.info(
             "APNs: Removing invalid/expired token %s (%s)", device.token, result.description
         )
-        if isinstance(device, RemotePushDevice):
+        if settings.ZILENCER_ENABLED and isinstance(device, RemotePushDevice):
```

Comment:
This is not correct actually, this will lead to self-hosted servers unable to delete expired / invalid tokens for the E2EE codepath.

I'll post a fix shortly.

Reply:
Ah, no! this is correct.